src/main/java/com/cloudbees/jenkins/plugins/bitbucket/MergeWithGitSCMExtension.java

```diff
@@ -1,50 +0,0 @@
-/*
- * The MIT License
- *
- * Copyright (c) 2017, CloudBees, Inc.
- *
- * Permission is hereby granted, free of charge, to any person obtaining a copy
- * of this software and associated documentation files (the "Software"), to deal
- * in the Software without restriction, including without limitation the rights
- * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
- * copies of the Software, and to permit persons to whom the Software is
- * furnished to do so, subject to the following conditions:
- *
- * The above copyright notice and this permission notice shall be included in
- * all copies or substantial portions of the Software.
- *
- * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
- * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
- * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
- * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
- * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
- * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
- * THE SOFTWARE.
- */
-package com.cloudbees.jenkins.plugins.bitbucket;
-
-import edu.umd.cs.findbugs.annotations.CheckForNull;
-import edu.umd.cs.findbugs.annotations.NonNull;
-import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
-import java.io.ObjectStreamException;
-import org.kohsuke.accmod.Restricted;
-import org.kohsuke.accmod.restrictions.DoNotUse;
-
-/**
- * Retained for data migration.
- *
- * @deprecated use {@link jenkins.plugins.git.MergeWithGitSCMExtension}
- */
-@Deprecated
-@Restricted(DoNotUse.class)
-@SuppressFBWarnings("NM_SAME_SIMPLE_NAME_AS_SUPERCLASS")
-public class MergeWithGitSCMExtension extends jenkins.plugins.git.MergeWithGitSCMExtension {
-
-    MergeWithGitSCMExtension(@NonNull String baseName, @CheckForNull String baseHash) {
-        super(baseName, baseHash);
-    }
-
-    private Object readResolve() throws ObjectStreamException {
-        return new jenkins.plugins.git.MergeWithGitSCMExtension(getBaseName(), getBaseHash());
-    }
```

Comment:
MergeWithGitSCMExtension was deprecated in Bitbucket Branch Source 2.2.3 (28 Jul 2017) for [JENKINS-45771](https://issues.jenkins.io/browse/JENKINS-45771). If users have projects saved by older versions and have not saved them again in the meantime, then they still need this class for resolving the correct extension when Jenkins loads those projects. Perhaps that isn't very likely because it's been four years already. Or perhaps this would still deserve a mention in the release notes and [`<hpi.compatibleSinceVersion>2.2.3</hpi.compatibleSinceVersion>](https://www.jenkins.io/doc/developer/plugin-development/mark-a-plugin-incompatible/).

Reply:
MergeWithGitSCMExtension wasn't used in a project: I found no instance creation. Is there any "magic" behind i don't aware of? See https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/520/commits/1fe75e91b2dc9d4106c524b6870c5ba1409bdb4b.

Reply:
If you configure a multibranch project with the "Discover pull requests from origin" behaviour and the "Merging the pull request with the current target branch revision" strategy, and let it scan the repository once and find a pull request, then the `config.xml` file of the pull request in the Jenkins controller will contain something like this:

```XML
<?xml version='1.1' encoding='UTF-8'?>
<flow-definition plugin="workflow-job@2.42">
  <!-- ...omitted... -->
  <properties>
    <org.jenkinsci.plugins.workflow.multibranch.BranchJobProperty plugin="workflow-multibranch@2.26">
      <branch plugin="branch-api@2.7.0">
        <!-- ...omitted... -->
        <head class="com.cloudbees.jenkins.plugins.bitbucket.PullRequestSCMHead" plugin="cloudbees-bitbucket-branch-source@726.vb0c1ea6c9336">
          <!-- ...omitted... -->
          <strategy>MERGE</strategy>
        </head>
        <scm class="hudson.plugins.git.GitSCM" plugin="git@4.10.0">
          <configVersion>2</configVersion>
          <!-- ...omitted... -->
          <extensions>
            <!-- ...omitted... -->
            <jenkins.plugins.git.MergeWithGitSCMExtension>
              <baseName>remotes/upstream/master</baseName>
            </jenkins.plugins.git.MergeWithGitSCMExtension>
          </extensions>
        </scm>
        <!-- ...omitted... -->
      </branch>
    </org.jenkinsci.plugins.workflow.multibranch.BranchJobProperty>
    <!-- ...omitted... -->
  </properties>
  <!-- ...omitted... -->
</flow-definition>
```

If this config.xml had been saved by Bitbucket Branch Source 2.2.2 or earlier, I believe the tag would be `<com.cloudbees.jenkins.plugins.bitbucket.MergeWithGitSCMExtension>` rather than `<jenkins.plugins.git.MergeWithGitSCMExtension>`.  And then you'd need this deprecated class for loading the extension back from the file. The `readResolve` method would create an instance of `jenkins.plugins.git.MergeWithGitSCMExtension` so the XML would switch to that when next saved.

Reply:
Then i shall revert this :-) Sorry about extra work i produce... Still learning...

Reply:
Because this was deprecated so long ago and the compatibility issue would only affect pull requests, which aren't likely to stay open that long, this could still be OK to delete in a separate pull request. Even after deleting the deprecated class, the plugin could preserve compatibility by calling [XStream2.addCompatibilityAlias](https://javadoc.jenkins.io/hudson/util/XStream2.html#addCompatibilityAlias-java.lang.String-java.lang.Class-) like [job-dsl-plugin does](https://github.com/jenkinsci/job-dsl-plugin/blob/job-dsl-1.78/job-dsl-plugin/src/main/groovy/javaposse/jobdsl/plugin/DescriptorImpl.java#L115-L126).